lib/floki/finder.ex

```diff
@@ -95,7 +95,12 @@ defmodule Floki.Finder do
           nil -> [n|res_acc]
           _ ->
             {_, _, children_nodes} = n
-            traverse(children_nodes, sibling_nodes, selector.combinator.selector, res_acc)
+
+            if selector.combinator.match_type == :child do
```

Comment:
We much check if the next element is worth "diving into". If it is – continue the "child selector combo" by calling just `traverse_child(children_nodes...)`.

If the combinator is a different selector – this means the combo is successfully over (e.g. entire chil chain was resolved), and we can go back to normal traversal via `traverse(nodes...)`.